flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestMonitorSource.java

```diff
@@ -0,0 +1,344 @@
+  private static final TableChange EMPTY_EVENT = new TableChange(0, 0, 0L, 0L, 0);
+  private static final RateLimiterStrategy HIGH_RATE = RateLimiterStrategy.perSecond(100.0);
+  private static final RateLimiterStrategy LOW_RATE = RateLimiterStrategy.perSecond(1.0 / 10000.0);
+
+  @TempDir private File checkpointDir;
+
+  @ParameterizedTest
+  @ValueSource(booleans = {true, false})
+  void testChangeReaderIterator(boolean withDelete) {
+    if (withDelete) {
+      sql.exec(
+          "CREATE TABLE %s (id int, data varchar, PRIMARY KEY(`id`) NOT ENFORCED) WITH ('format-version'='2', 'write.upsert.enabled'='true')",
+          TABLE_NAME);
+    } else {
+      sql.exec("CREATE TABLE %s (id int, data varchar)", TABLE_NAME);
+    }
+
+    TableLoader tableLoader = sql.tableLoader(TABLE_NAME);
+    tableLoader.open();
+    Table table = tableLoader.loadTable();
+
+    MonitorSource.SchedulerEventIterator iterator =
+        new MonitorSource.SchedulerEventIterator(tableLoader, null, Long.MAX_VALUE);
+
+    // For an empty table we get an empty result
+    assertThat(iterator.next()).isEqualTo(EMPTY_EVENT);
+
+    // Add a single commit and get back the commit data in the event
+    sql.exec("INSERT INTO %s VALUES (1, 'a')", TABLE_NAME);
+    table.refresh();
+    long dataSize =
+        table.currentSnapshot().addedDataFiles(table.io()).iterator().next().fileSizeInBytes();
+    long deleteSize =
+        withDelete
+            ? table
+                .currentSnapshot()
+                .addedDeleteFiles(table.io())
+                .iterator()
+                .next()
+                .fileSizeInBytes()
+            : 0;
+    assertThat(iterator.next())
+        .isEqualTo(new TableChange(1, withDelete ? 1 : 0, dataSize, deleteSize, 1));
+    // Make sure that consecutive calls do not return the data again
+    assertThat(iterator.next()).isEqualTo(EMPTY_EVENT);
+
+    // Add two more commits, but fetch the data in one loop
+    sql.exec("INSERT INTO %s VALUES (2, 'b')", TABLE_NAME);
+    table.refresh();
+    dataSize =
+        table.currentSnapshot().addedDataFiles(table.io()).iterator().next().fileSizeInBytes();
+    deleteSize =
+        withDelete
+            ? table
+                .currentSnapshot()
+                .addedDeleteFiles(table.io())
+                .iterator()
+                .next()
+                .fileSizeInBytes()
+            : 0;
+    sql.exec("INSERT INTO %s VALUES (3, 'c')", TABLE_NAME);
+    table.refresh();
+    dataSize +=
+        table.currentSnapshot().addedDataFiles(table.io()).iterator().next().fileSizeInBytes();
+    deleteSize +=
+        withDelete
+            ? table
+                .currentSnapshot()
+                .addedDeleteFiles(table.io())
+                .iterator()
+                .next()
+                .fileSizeInBytes()
+            : 0;
+    assertThat(iterator.next())
+        .isEqualTo(new TableChange(2, withDelete ? 2 : 0, dataSize, deleteSize, 2));
```

Comment:
nit: tied to some previous comment. it is a bit hard to read/remember what each arg is for (and they are all ints or longs).